src/components/ScrollTrack/ScrollTrack.js

```diff
@@ -142,21 +159,66 @@ class ScrollTrack extends Component {
     const { parentWidth, trackWidth } = this.getNodeWidths()
     let nextForward = this.state.left - parentWidth
     const fullForward = parentWidth - trackWidth
+    const { onBeforeNext, onAfterNext } = this.props
 
     // already is, or is going to be, full forward
     if (nextForward <= fullForward) { nextForward = fullForward }
 
-    this.setState({ left: nextForward }, this.computeSlideAttributes)
+    const callbackProps = {
+      atStart: trackWidth <= parentWidth,
+      atEnd: fullForward === nextForward,
+      slideTo: nextForward,
+      parentWidth,
+      trackWidth
+    }
+
+    if (!onBeforeNext) {
+      this.updateLeftValue({
+        left: nextForward,
+        callback: onAfterNext,
+        callbackProps
+      })
+    } else {
+      onBeforeNext(callbackProps).then(() => {
```

Comment:
Seems really clean to me. Like we talked about yesterday for a bit, the only other ways i can see to do this is either 1) let rendering components handle pagination 2) have ScrollTrack pass a callback back to it's parent 3) expose a ref on ScrollTrack and have the parent call a method directly to move the scroll position. None of which sound all that exciting :)